packages/rOpenScPCA/R/calculate-clusters.R

```diff
@@ -27,36 +33,61 @@
 #'
 #' @examples
 #' \dontrun{
-#' # cluster using default parameters
+#' # cluster PCs from a SingleCellExperiment object using default parameters
+#' cluster_df <- calculate_clusters(sce_object)
+#'
+#' # cluster PCs from a Seurat object using default parameters
+#' cluster_df <- calculate_clusters(seurat_object)
+#'
+#' # cluster directly from a matrix using default parameters
 #' cluster_df <- calculate_clusters(pca_matrix)
 #'
-#' # cluster using the leiden algorithm with a resolution of 0.1
+#' # cluster directly from a matrix using the leiden algorithm with a resolution of 0.1
 #' cluster_df <- calculate_clusters(pca_matrix, algorithm = "leiden", resolution = 0.1)
 #'
-#' # cluster using the leiden algorithm with a non-default of 3 iterations
+#' # cluster directly from a matrix using the leiden algorithm with 3 iterations
 #' cluster_df <- calculate_clusters(
 #'   pca_matrix,
 #'   algorithm = "leiden",
 #'   cluster_args = list(n_iterations = 3)
 #' )
 #' }
 calculate_clusters <- function(
-    mat,
+    x,
     algorithm = c("louvain", "walktrap", "leiden"),
     weighting = c("jaccard", "rank", "number"),
     nn = 10,
     resolution = 1, # louvain or leiden
     objective_function = c("CPM", "modularity"), # leiden only
     cluster_args = list(),
-    seed = NULL) {
+    seed = NULL,
+    pc_name = NULL) {
   if (!is.null(seed)) {
     set.seed(seed)
   }
 
+  # check and prepare matrix, as needed
+  if (any(class(x) %in% c("matrix", "Matrix"))) {
+    stopifnot(
+      "The matrix must have row names representing cell ids, e.g. barcodes." = is.character(rownames(x))
+    )
+    pca_matrix <- x # redefine with better variable name
+  } else {
+    if (is(x, "SingleCellExperiment")) {
+      pc_name <- ifelse(is.null(pc_name), "PCA", pc_name)
+      object_type <- "sce"
+    } else if (is(x, "Seurat")) {
+      pc_name <- ifelse(is.null(pc_name), "pca", pc_name)
+      object_type <- "seurat"
+    } else {
+      stop("The first argument should be one of: a SingleCellExperiment object, a Seurat object, or a matrix with row names.")
+    }
+    pca_matrix <- extract_pc_matrix(x, object_type, pc_name)
+  }
```

Comment:
Is there any reason this would not work with the previous version of the function?
```suggestion
  } else if (is(x, "SingleCellExperiment") || is(x, "Seurat")) {
    pca_matrix <- extract_pc_matrix(x, pc_name)
  } else {
    stop("The first argument should be one of: a SingleCellExperiment object, a Seurat object, or a matrix with row names.")
  }
```